### DISCUSSION TITLE
[Uniques v2] Basic marketplace functions

### DISCUSSION THREAD

**Discussion by jsidorenko** [paritytech/substrate#11794] [Ideas]
## Idea:
In a classic marketplace implementation, a user needs to provide full access to his NFT to a 3rd-party app (marketplace). This approach usually leads to well-known security problems where a bug in the marketplace’s codebase (smart-contract/backend or just the way the private keys are handled) allows an attacker to steal all the assets.

As a possible solution, we could add 2 new methods to our Uniques pallet: 
1. `set_seller(collection_id, item_id, seller, price, tips)` - where the user could set a marketplace that would be able to sell his NFT
2. `buy_from_seller(collection_id, item_id, price, tips)` 

\* The `tips` param would be an array of addresses and amounts to send the tokens.

With that approach, any marketplace would be able to display the trusted items to sell, while having no control over the asset itself.

**Comment by Szegoo** [comment]
When setting a marketplace for an item we need to lock that item up, so we should have an additional method for removing an item from a marketplace so that the item becomes transferable again.

So we should also have something like this: ```remove_from_seller(collection_id, item_id)``` - only callable by the item owner.

**Reply by gilescope** [reply]
Can we have seller be sellers please? Competition is a good thing. No reason why opensea and looksrare can't both try and sell my nft.

**Reply by jsidorenko** [reply]
It means for opensea I would set one array of tips and for looksrare another, right? In that case, by reading the on-chain data anyone could simply execute the set of tips he likes the most and the UI won't be able to control that

**Reply by gilescope** [reply]
I think it's ok for sellers to have different rates. Price is only one factor when deciding who to buy from. OpenSea isn't the cheapest platform out there but still they come... And yes smart wallets could figure out that you might be able to buy it cheaper elsewhere.

**Reply by gilescope** [reply]
Hmm maybe add_seller and remove_seller would be nicer as you probably would not want to add multiple sellers in the same transaction...

**Reply by jsidorenko** [reply]
I agree with the `add_seller` and `remove_seller`, it's better to have 2 separate methods. I've implemented them here: https://github.com/paritytech/substrate/pull/12189

**Reply by jsidorenko** [reply]
Regarding the multiple sellers.. In the EVM world, you usually give full control over your NFT to some seller (e.g. OpenSea), which controls the whole flow (e.g. royalties). You won't be able to put your NFT on another marketplace at the same time. Additionally, you would need to trust that seller's contracts, otherwise, you could completely lose your asset.
We're trying to solve that by having that `add_seller` method with the tips hardcoded. You can imagine that as if you decide to sell on OpenSea, and they charge 3% in fees, then you would call `add_seller(OpenSea, your_price, 3%)`. OpenSea platform would be able to read all the listed NFTs from some new storage mapping (SellerItems) and display them on their UI if the tips are correct.

The main difference between that approach and the optional tips in Buy&Sell is that here you have the control over the tips that will be paid. In reality, those tips will be an array of records, like 1.5% to the marketplace, 1% to the original creator, etc..
